File: SymmetricBT.py
# time: O(n)
# space : O(h)
# Did this code successfully run on Leetcode : Yes

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        return self.isSym(root, root)

    def isSym(self, tn1, tn2):

        # base cond
        if tn1 is None and tn2 is None:
            return True
        if tn1 is None or tn2 is None:
            return False

        # logic
        return (tn1.val == tn2.val) \
               and (self.isSym(tn1.right, tn2.left)) \
               and (self.isSym(tn1.left, tn2.right))

File: SymmetricBT.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
//TimeComplexity:O(N) Traversing each node
//SpaceComplexity:O(N) Recurrsive Stack
class Solution {
    public boolean isSymmetric(TreeNode root) {
        return isSymmetric(root,root);
    }
    public boolean isSymmetric(TreeNode t1, TreeNode t2) {
        if(t1==null && t2==null)
            return true;
        if(t1==null || t2==null)
            return false;
        return (t1.val==t2.val)
                && isSymmetric(t1.left,t2.right)
                && isSymmetric(t2.left,t1.right);
    }
}